libraries/chain/include/eosio/chain/symbol.hpp

```diff
@@ -0,0 +1,151 @@
+/**
+ *  @file
+ *  @copyright defined in eos/LICENSE.txt
+ */
+#pragma once
+#include <fc/exception/exception.hpp>
+#include <eosio/chain/types.hpp>
+#include <string>
+
+namespace eosio {
+   namespace chain {
+      
+      static constexpr uint64_t string_to_symbol_c(uint8_t precision, const char* str) {
+         uint32_t len = 0;
+         while (str[len]) ++len;
+         
+         uint64_t result = 0;
+         // No validation is done at compile time
+         for (uint32_t i = 0; i < len; ++i) {
+            result |= (uint64_t(str[i]) << (8*(1+i)));
+         }
+
+         result |= uint64_t(precision);
+         return result;
+      }
+      
+#define SY(P,X) ::eosio::chain::string_to_symbol_c(P,#X)
+
+      static uint64_t string_to_symbol(uint8_t precision, const char* str) {
+         try {
+            uint32_t len = 0;
+            while(str[len]) ++len;
+            uint64_t result = 0;
+            for (uint32_t i = 0; i < len; ++i) {
+               // All characters must be upper case alaphabets
+               FC_ASSERT (str[i] >= 'A' && str[i] <= 'Z', "invalid character in symbol name");
+               result |= (uint64_t(str[i]) << (8*(i+1)));
+            }
+            result |= uint64_t(precision);
+            return result;
+         } FC_CAPTURE_LOG_AND_RETHROW((str))
+      }
+
+      class symbol {
+         public:
+            symbol(uint8_t p, const char* s): m_value(string_to_symbol(p, s)) { }
+            symbol(uint64_t v = SY(4, EOS)): m_value(v) { }
+            static symbol from_string(const string& from)
+            {
+               try {
+                  string s = fc::trim(from);
+                  if (s.empty()) {
+                     return SY(4, EOS);
+                  }
+                  auto comma_pos = s.find(',');
+                  if (comma_pos == string::npos) {
+                     return string_to_symbol(4, s.c_str());
+                  }
+                  if (comma_pos == 0) {
+                     return string_to_symbol(4, s.substr(1).c_str());
+                  }
+                  auto prec_part = s.substr(0, comma_pos);
+                  uint8_t p = fc::to_int64(prec_part);
+                  string name_part = s.substr(comma_pos + 1);
+                  return string_to_symbol(p, name_part.c_str());
+               } FC_CAPTURE_LOG_AND_RETHROW((from))
+            }
+            uint64_t value() const { return m_value; }
+            operator uint64_t() const { return m_value; }
+            bool valid() const
+            { 
+               const auto& s = name();
+               return all_of(s.begin(), s.end(), [](char c)->bool { return (c >= 'A' && c <= 'Z'); });
```

Comment:
Can this just call `return valid_name(s)`

Reply:
Yes. Changed that.